core/src/main/scala/filodb.core/memstore/PartKeyIndex.scala

```diff
@@ -205,3 +419,209 @@ trait PartKeyIndexDownsampled extends PartKeyIndexRaw {
                                    endTime: Long, func: (BytesRef) => Unit): Int
 
 }
+
+/**
+ * Base class to convert incoming FiloDB ColumnFilters into the index native
+ * query format.  Uses a visitor pattern to allow the implementation to build
+ * different query object patterns.
+ *
+ * The query language supports the following:
+ * * Boolean queries (many subqueries with MUST / MUST_NOT).  These may be nested.
+ * * Equals queries (term match)
+ * * Regex queries (regex term match)
+ * * TermIn queries (term presence in list)
+ * * Prefix queries (term starts with)
+ * * MatchAll queries (match all values)
+ * * Range queries (start to end on a long value)
+ *
+ * This list may expand over time.
+ *
+ * The top level is always a Boolean query, even if it only has one child query.  A
+ * Boolean query will never have zero child queries.  Leaf nodes are always non-Boolean
+ * queries.
+ *
+ * For example:
+ *   + represents MUST, - represents MUST_NOT
+ *   Input query -> ((+Equals(Col, A) -Equals(Col, B)) +Regex(Col2, C.*))
+ *
+ *   The visitor call pattern would be:
+ *    * visitStartBooleanQuery
+ *    * visitStartBooleanQuery
+ *    * visitEqualsQuery(Col, A, MUST)
+ *    * visitEqualsQuery(Col, B, MUST_NOT)
+ *    * visitEndBooleanQuery
+ *    * visitRegexQuery(Col2, C.*, MUST)
+ *    * visitEndBooleanQuery
+ *
+ * This class should be used instead of putting query parsing code in each
+ * implementation so that common optimizations, such as query rewriting,
+ * can occur in one place.
+ */
+abstract class PartKeyQueryBuilder {
```

Comment:
This is a new pattern for existing code.  The idea is to separate query optimization (rewriting regexes, etc) from conversation into index specific query classes.  We do this with a visitor pattern, each implementation gets calls for logical optimized query types and it uses that to build an index query data structure.